schema/pgsql/schema.sql

```diff
@@ -0,0 +1,1912 @@
+-- Icinga DB | (c) 2021 Icinga GmbH | GPLv2+
+
+CREATE DOMAIN bytea20 AS bytea CONSTRAINT exactly_20_bytes_long CHECK ( VALUE IS NULL OR octet_length(VALUE) = 20 );
+CREATE DOMAIN bytea16 AS bytea CONSTRAINT exactly_16_bytes_long CHECK ( VALUE IS NULL OR octet_length(VALUE) = 16 );
+CREATE DOMAIN bytea4 AS bytea CONSTRAINT exactly_4_bytes_long CHECK ( VALUE IS NULL OR octet_length(VALUE) = 4 );
+
+CREATE DOMAIN biguint AS bigint CONSTRAINT positive CHECK ( VALUE IS NULL OR 0 <= VALUE );
+CREATE DOMAIN uint AS bigint CONSTRAINT between_0_and_4294967295 CHECK ( VALUE IS NULL OR VALUE BETWEEN 0 AND 4294967295 );
+CREATE DOMAIN smalluint AS int CONSTRAINT between_0_and_65535 CHECK ( VALUE IS NULL OR VALUE BETWEEN 0 AND 65535 );
+CREATE DOMAIN tinyuint AS smallint CONSTRAINT between_0_and_255 CHECK ( VALUE IS NULL OR VALUE BETWEEN 0 AND 255 );
+
+CREATE TYPE boolenum_t AS ENUM ( 'n', 'y' );
```

Comment:
```
2021-12-06T16:41:09.340+0100	FATAL	icingadb	pq: operator does not exist: boolenum = unknown
can't perform "SELECT id, heartbeat FROM icingadb_instance WHERE environment_id = $1 AND responsible = $2 AND id <> $3 AND heartbeat > $4"
```

Looks like you have to do `WHERE responsible::boolenum_t = 'y'` in queries now, so not sure if this is indeed a good idea.

Reply:
You mean `WHERE responsible = 'y'::boolenum_t`?

Reply:
No? (Don't ask me why it's that way)

```
icingadb=# SELECT COUNT(*) FROM icingadb_instance WHERE responsible::boolenum_t = 'y';
 count 
-------
     0
(1 row)

icingadb=# SELECT COUNT(*) FROM icingadb_instance WHERE responsible = 'y'::boolenum_t;
ERROR:  operator does not exist: boolenum = boolenum_t
LINE 1: ...COUNT(*) FROM icingadb_instance WHERE responsible = 'y'::boo...
                                                             ^
HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
```

Reply:
Second line is probably a hint why it doesn't work:
```
icingadb=# SELECT COUNT(*) FROM icingadb_instance WHERE responsible = 'y'::boolenum;
ERROR:  operator does not exist: boolenum = boolenum
LINE 1: ...COUNT(*) FROM icingadb_instance WHERE responsible = 'y'::boo...
                                                             ^
HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
```

Reply:
And w/o _t?

Reply:
`s/responsible/responsible::boolenum_t/` in Go or ENUM -> CHAR(1) CHECK...?

Reply:
> `s/responsible/responsible::boolenum_t/` in Go

No way, that would defeat the whole purpose of trying to behave somewhat like MySQL. Web would have to do the same for each query. Then we could just use the standard boolean type with less impact.

Reply:
https://github.com/Icinga/icingadb/pull/221#discussion_r759395335

Reply:
As we're already talking about schema equality and sort order.. what if we explicitly specify the DEFAULTs for both RDBMS', drop the boolenum DOMAIN in favour of the boolenum_t TYPE and.. it will just work?

Reply:
We don't have to change anything for MySQL:

> If an ENUM column is declared to permit NULL, the NULL value is a valid value for the column, and the default value is NULL. If an ENUM column is declared NOT NULL, its default value is the first element of the list of permitted values.

> ENUM values are sorted based on their index numbers, which depend on the order in which the enumeration members were listed in the column specification

https://dev.mysql.com/doc/refman/8.0/en/enum.html

Reply:
OK.. so I'd just do https://github.com/Icinga/icingadb/pull/221#discussion_r765670379 w/o touching the MySQL schema.. ?

Reply:
I don't know. The sort order has not yet been clarified. You should find out what we should do.

Reply:
1. "The ordering of the values in an enum type is the order in which the values were listed when the type was created." – https://www.postgresql.org/docs/9.1/datatype-enum.html
2. I've just suggested what we should do. To be precise: compared to status quo (DOMAINs) I'd like to make the columns of just the ENUM types and explicitly specify a default per column.

Reply:
Yes, I know, schema equality.. but in this case we'd provide schema equality – effective defaults – by having schema unequality – explicitly (not) specified defaults.

Reply:
We don't have `DOMAIN` for the bools in PostgreSQL at the moment. We have `ENUM`. And the problem with that is that it requires casting, as far as I understand @julianbrost correctly.

Reply:
That's not true:

https://github.com/Icinga/icingadb/blob/74d8bb9a171d31e79e372d31128ebd1aebc768fe/schema/pgsql/schema.sql#L12-L26

https://github.com/Icinga/icingadb/blob/74d8bb9a171d31e79e372d31128ebd1aebc768fe/schema/pgsql/schema.sql#L55-L58

Reply:
> And the problem with that is that it requires casting, as far as I understand @julianbrost correctly.

I don't know in detail if there's a way to fix it but on PostgreSQL 14 with the current enum + type + domain construction, I wasn't able to do any queries without casting.

Reply:
OK, let's formulate the question as unambiguous as possible:

If I'd do `CREATE TYPE boolenum AS ENUM ( 'n', 'y' );` and `active_checks_enabled boolenum NOT NULL DEFAULT 'n',` for all columns and it would work, would you accept it or not?

Reply:
Damn. Seems its a s/bug/feature/: https://www.postgresql.org/message-id/2971.1294769104%40sss.pgh.pa.us

Reply:
What about this (for all of our domains)?

```
idb=# SELECT COUNT(*) FROM icingadb_instance WHERE responsible = 'y';
ERROR:  operator does not exist: boolenum = unknown
LINE 1: ...ECT COUNT(*) FROM icingadb_instance WHERE responsible = 'y';
                                                                 ^
HINT:  No operator matches the given name and argument type(s). You might need to add explicit type casts.
idb=# CREATE FUNCTION compare(boolenum, unknown) RETURNS boolean AS 'SELECT $1::boolenum_t = $2::char::boolenum_t;' LANGUAGE SQL IMMUTABLE RETURNS NULL ON NULL INPUT;
CREATE FUNCTION
idb=# CREATE OPERATOR = (LEFTARG = boolenum, RIGHTARG = unknown, PROCEDURE = compare);
CREATE OPERATOR
idb=# SELECT COUNT(*) FROM icingadb_instance WHERE responsible = 'y';
 count
-------
     0
(1 row)

idb=#
```

@lippserd @nilmerg Is web using more operators?

Reply:
Does `ORDER BY` work?

Reply:
But I wouldn't be surprised if it needed an operator `<`. Have you tested it?

Anyways, I'm not convinced of this idea. Your question shouldn't be if web uses more operators but rather also if web or anything else might want to use other operators in the future. And if we don't get that answer right now, we'd have to upgrade the schema and add operators. So I think just adding `DEFAULT 'n'` to the column definitions is the easier way.

Reply:
Surprisingly the non-ENUM domains just work:

```
idb=# \x
Expanded display is on.
idb=# select * from history order by event_time;
(0 rows)

idb=# select * from history where event_time > 0 order by event_time;
(0 rows)

idb=# select * from history where event_time = 0 order by event_time;
(0 rows)

idb=# select * from host where id = 'x';
(0 rows)

idb=#
```